coding/src/zoda/mod.rs

```diff
@@ -0,0 +1,793 @@
+    /// of recovery shards for each lane.
+    fn encode_lanes<'f>(
+        lanes: impl IndexedParallelIterator<Item = impl Iterator<Item = &'f F>>,
+        shard_size: usize,
+        rate_inverse: usize,
+    ) -> Result<Vec<Vec<F>>, RSError>
+    where
+        F: 'f,
+    {
+        lanes
+            .map(|mut lane| {
+                // Add each lane element as an original shard. (RS rate: rate_inverse^-1)
+                let mut encoder = ReedSolomonEncoder::new(
+                    shard_size,
+                    shard_size * rate_inverse.saturating_sub(1),
+                    F::BYTE_SIZE,
+                )?;
+                lane.try_for_each(|el| encoder.add_original_shard(el.to_le_bytes()))?;
+
+                // Encode the lane and return the parity shards.
+                let result = encoder.encode()?;
+                let recovery_shards = result
+                    .recovery_iter()
+                    .map(F::from_le_bytes)
+                    .collect::<Vec<_>>();
+                Ok(recovery_shards)
+            })
+            .collect::<Result<Vec<_>, RSError>>()
+    }
+
+    /// Creates a [`Tree`] from the given elements.
+    fn merkleize<E: Encode>(
+        leaves: impl Iterator<Item = E>,
+        leaf_count: usize,
+        hasher: &mut H,
+    ) -> Tree<H> {
+        let mut tree = TreeBuilder::<H>::new(leaf_count);
+
+        // Include each encoded item's hash as a leaf in the tree.
+        leaves.for_each(|v| {
+            hasher.update(v.encode().as_ref());
+            tree.add(&hasher.finalize());
+        });
+
+        tree.build()
+    }
+}
+
+/// A proof that the encoding for a certain [`Commitment`] was performed correctly.
+#[derive(Debug, Clone)]
+pub struct EncodingProof<F: BinaryField> {
+    y_r: Vec<F>,
+    w_r_prime: Vec<F>,
+}
+
+impl<F: BinaryField> Write for EncodingProof<F> {
+    fn write(&self, buf: &mut impl bytes::BufMut) {
+        self.y_r.write(buf);
+        self.w_r_prime.write(buf);
+    }
+}
+
+impl<F: BinaryField> EncodeSize for EncodingProof<F> {
+    fn encode_size(&self) -> usize {
+        self.y_r.encode_size() + self.w_r_prime.encode_size()
+    }
+}
+
+impl<F: BinaryField<Cfg = ()>> Read for EncodingProof<F> {
+    type Cfg = ();
+
+    fn read_cfg(buf: &mut impl bytes::Buf, _: &Self::Cfg) -> Result<Self, commonware_codec::Error> {
+        let range_cfg = RangeCfg::from(0..=usize::MAX);
+
+        let y_r = Vec::<F>::read_cfg(buf, &(range_cfg, ()))?;
+        let w_r_prime = Vec::<F>::read_cfg(buf, &(range_cfg, ()))?;
+
+        Ok(Self { y_r, w_r_prime })
+    }
+}
+
+/// A set of openings for sampled rows and columns in a [`Commitment`].
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Sample<H: Hasher, F: BinaryField> {
+    row_root: H::Digest,
+    col_root: H::Digest,
+    rows: HashMap<usize, (Vec<F>, TreeProof<H>)>,
+    cols: HashMap<usize, (Vec<F>, TreeProof<H>)>,
+}
```

Comment:
It doesn't make sense for `Sample` to also be an authority for which rows should have been sampled. The adversary should generate the sample, based on the rows and columns the verifier chooses. But, crucially, the verifier needs to not just take the sample's word for which columns and rows were sampled, but actually check that these are the same. A simpler way to do this is for the `Sample` to not contain references to which rows these are, but instead for the verify function to know this, and the sample to just have them in the right order.